lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -718,3 +726,111 @@ instance
                 addresses (liftPaymentAddress @n @k) (externalPool s)
         in
             nonChangeAddresses <> changeAddresses
+
+--------------------------------------------------------------------------------
+--
+-- SeqAnyState
+--
+-- For benchmarking and testing arbitrary large sequential wallets.
+
+-- | An "unsound" alternative that can be used for benchmarking and stress
+-- testing. It re-uses the same underlying structure as the `SeqState` but
+-- it discover addresses based on an arbitrary ratio instead of decrypting the
+-- derivation path.
+--
+-- The proportion is stored as a type-level parameter so that we don't have to
+-- alter the database schema to store it. It simply exists and depends on the
+-- caller creating the wallet to define it.
+newtype SeqAnyState (network :: NetworkDiscriminant) key (p :: Nat) = SeqAnyState
+    { innerState :: SeqState network key
+    } deriving (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    , Show (k 'AddressK XPub)
+    , Show (KeyFingerprint "payment" k)
+    ) => Show (SeqAnyState n k p)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    , NFData (k 'AddressK XPub)
+    , NFData (KeyFingerprint "payment" k)
+    )
+    => NFData (SeqAnyState n k p)
+
+-- | Initialize the HD random address discovery state from a root key and RNG
+-- seed.
+--
+-- The first argument is expected to be a ratio (between 0 and 1) of addresses
+-- we ought to simply recognize as ours. So, giving .5 means that 50% of the
+-- entire address space of the network will be considered ours, picked randomly.
+mkSeqAnyState
+    :: forall (p :: Nat) n k.
+        ( SoftDerivation k
+        , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+        , MkKeyFingerprint k Address
+        , WalletKey k
+        , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+        )
+    => (k 'RootK XPrv, Passphrase "encryption")
+    -> AddressPoolGap
+    -> SeqAnyState n k p
+mkSeqAnyState credentials poolGap = SeqAnyState
+    { innerState = mkSeqStateFromRootXPrv credentials poolGap
+    }
+
+instance
+    ( SoftDerivation k
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+    , KnownNat p
+    ) => IsOurs (SeqAnyState n k p) Address
+  where
+    isOurs (Address bytes) st@(SeqAnyState inner)
+        | crc32 bytes < p =
+            let
+                edge = Map.size (indexedKeys $ externalPool inner)
+                ix = toEnum (edge - fromEnum (gap $ externalPool inner))
```

Comment:
Always the maximum gap between addresses, causing us to have to generate a full `gap` amount of addresses for each discovery? I guess this might not substantially affect the results though?

Reply:
This is actually the opposite, it uses the first address of the edge, which results in only a single new address to be generated, so each discovery creates a new address which should be roughly equivalent to a standard use-case where addresses aren't re-use and when looked from the entire run: in practice, we may sometimes generate no new addresses and sometimes have to generate a full gap, but for a number of addresses N, the number of total generations is the same regardless of whether they were generated one by one or, by batches (arguably, there's a little difference? but :shrug:)